engine/src/main/java/org/terasology/rendering/nui/editor/NUIEditorContextMenuBuilder.java

```diff
@@ -0,0 +1,350 @@
+            } else {
+                populateGenericContextMenu(node, addLevel);
+            }
+        }
+    }
+
+    private void populateWidgetContextMenu(JsonTree node, ContextMenuLevel addLevel) {
+        // Get the ClassMetadata of the widget with the given type.
+        String type = (String) node.getChildWithKey("type").getValue().getValue();
+        ClassMetadata<? extends UIWidget, ?> elementMetadata = nuiManager
+            .getWidgetMetadataLibrary()
+            .resolve(type, ModuleContext.getContext());
+
+        if (elementMetadata != null) {
+            for (FieldMetadata fieldMetadata : elementMetadata.getFields()) {
+                Field field = fieldMetadata.getField();
+                field.setAccessible(true);
+
+                // Add options for all @LayoutConfig-annotated fields.
+                if (field.isAnnotationPresent(LayoutConfig.class)) {
+                    String name = getNodeName(field);
+                    if (!node.hasChildWithKey(name)) {
+                        addLevel.addOption(name, n -> {
+                            try {
+                                createChild(name, node, field, elementMetadata);
+                            } catch (IllegalAccessException e) {
+                                return;
+                            }
+                            addContextMenuListeners.forEach(UpdateListener::onAction);
+                        }, node, true);
+                    }
+                }
+            }
+        } else {
+            logger.warn("Unknown widget type: {}", type);
+        }
+    }
+
+    private void populateLayoutContextMenu(JsonTree node, ContextMenuLevel addLevel) {
+        String type = "";
+        JsonTree currentNode = node;
+        ClassMetadata layoutMetadata = null;
+        Class layoutHintType = null;
+
+        // Iterate down the tree and attempt to find a UILayout node.
+        while (layoutHintType == null) {
+            if (currentNode.hasSiblingWithKey("type")) {
+                type = (String) currentNode.getSiblingWithKey("type").getValue().getValue();
+                layoutMetadata = nuiManager.getWidgetMetadataLibrary().resolve(type, ModuleContext.getContext());
+
+                // If one is found, attempt to get the layout hint type.
+                try {
+                    layoutHintType = (Class) ReflectionUtil
+                        .getTypeParameter(layoutMetadata.getType().getGenericSuperclass(), 0);
+                } catch (NullPointerException e) {
+                    layoutHintType = null;
+                }
+            }
+            if (layoutHintType == null) {
+                if (currentNode.isRoot()) {
+                    break;
+                }
+                currentNode = (JsonTree) currentNode.getParent();
+            }
+        }
+
+        if (layoutHintType != null) {
+            for (Field field : layoutHintType.getDeclaredFields()) {
```

Comment:
I wonder if that is the right way of doing it, as it is not the terasology way of getting field meta data.

That the field are annotated with FieldMetadata is one of the few classes that evaluates the SerializedName annotation. So I would assume that during widget creation the FieldMetadata gets also used as source of information about widget fields.